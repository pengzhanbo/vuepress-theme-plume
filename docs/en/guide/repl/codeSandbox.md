---
title: Code Sandbox
icon: lucide:codesandbox
createTime: 2025/10/08 03:42:13
permalink: /en/guide/code/code-sandbox/
---

The theme supports embedding [Code Sandbox](https://codesandbox.io) in Markdown files.

## Configuration

This feature is disabled by default. You can enable it in the configuration file.

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      codesandbox: true, // [!code highlight]
    },
  })
})
```

## Syntax

### Basic Syntax

Embed Code Sandbox in the page:

```md
@[codesandbox](id)
```

Use Code Sandbox redirect button:

```md
@[codesandbox button](workspace/id)
```

### Extended Options

```md
@[codesanbox title="xxx" layout="Editor+Preview" height="500px" navbar="false" console](id#filepath)
```

- `id`: Code Sandbox identifier
- `title`: Code Sandbox title
- `layout`: Code preview layout, options: `Preview`, `Editor`, `Editor+Preview`
- `height`: Code preview height
- `navbar`: Whether to display navigation bar, defaults to true
- `console`: Whether to display console, defaults to false
- `filepath`: File path

## Examples

Code Sandbox redirect button:

```md
@[codesandbox button](reaction/5wyzu)
```

Output:

@[codesandbox button](reaction/5wyzu)

Code Sandbox embedded in page:

```md
@[codesandbox](5wyzu)
```

Output:

@[codesandbox](5wyzu)
