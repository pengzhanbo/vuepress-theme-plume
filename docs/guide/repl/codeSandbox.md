---
title: Code Sandbox
icon: lucide:codesandbox
createTime: 2024/04/04 03:42:13
permalink: /guide/code/code-sandbox/
---

主题支持在 Markdown 文件中嵌入 [Code Sandbox](https://codesandbox.io)。

## 配置

此功能默认不启用，你可以在配置文件中启用它。

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      codesandbox: true, // [!code highlight]
    },
  })
})
```

## 语法

### 简单语法

将 Code Sandbox 嵌入到页面中

```md
@[codesandbox](id)
```

使用 Code Sandbox 跳转按钮

```md
@[codesandbox button](workspace/id)
```

### 更多选项

```md
@[codesanbox title="xxx" layout="Editor+Preview" height="500px" navbar="false" console](id#filepath)
```

- `id`: Code Sandbox ID
- `title`: Code Sandbox 标题
- `layout`: 代码预览布局 可选值： `Preview`， `Editor`， `Editor+Preview`
- `height`: 代码预览高度
- `navbar`: 是否显示导航栏，默认为 true
- `console`: 是否显示控制台，默认为 false
- `filepath`: 文件路径

## 示例

codeSandbox 跳转按钮：

```md
@[codesandbox button](reaction/5wyzu)
```

输出：

@[codesandbox button](reaction/5wyzu)

codeSandbox 内嵌到页面中：

```md
@[codesandbox](5wyzu)
```

输出：

@[codesandbox](5wyzu)
