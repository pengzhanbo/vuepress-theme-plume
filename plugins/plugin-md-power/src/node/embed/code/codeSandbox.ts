/**
 * @[codesandbox](id)
 * @[codesandbox share](user/id)
 * @[codesanbox title="xxx" layout="Editor+Preview" height="500px" navbar="false" console="false"](id#filepath)
 */
import type { PluginWithOptions } from 'markdown-it'
import type { CodeSandboxTokenMeta } from '../../../shared/index.js'
import { parseRect } from '../../utils/parseRect.js'
import { resolveAttrs } from '../../utils/resolveAttrs.js'
import { createEmbedRuleBlock } from '../createEmbedRuleBlock.js'

export const codeSandboxPlugin: PluginWithOptions<never> = (md) => {
  createEmbedRuleBlock<CodeSandboxTokenMeta>(md, {
    type: 'codesandbox',
    syntaxPattern: /^@\[codesandbox(?:\s+(embed|button))?([^\]]*)\]\(([^)]*)\)/,
    meta([, type, info = '', source = '']) {
      const { attrs } = resolveAttrs(info)
      const [profile, filepath = ''] = source.split('#')
      const [user, id] = profile.includes('/') ? profile.split('/') : ['', profile]

      return {
        width: attrs.width ? parseRect(attrs.width) : '100%',
        height: attrs.height ? parseRect(attrs.height) : '500px',
        user,
        id,
        title: attrs.title ?? '',
        console: attrs.console ?? false,
        navbar: attrs.navbar ?? true,
        layout: attrs.layout ?? '',
        type: (type || 'embed') as CodeSandboxTokenMeta['type'],
        filepath,
      }
    },
    content({ title, height, width, user, id, type, filepath, console, navbar, layout }) {
      return `<CodeSandboxViewer title="${title}" height="${height}" width="${width}" user="${user}" id="${id}" type="${type}" filepath="${filepath}" :console=${console} :navbar=${navbar} layout="${layout}" />`
    },
  })
}
