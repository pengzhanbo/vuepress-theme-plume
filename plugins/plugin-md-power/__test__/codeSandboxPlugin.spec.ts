import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'
import { codeSandboxPlugin } from '../src/node/embed/code/codeSandbox.js'

function createMarkdown() {
  return MarkdownIt().use((md) => {
    md.block.ruler.before('code', 'import_code', () => false)
    md.renderer.rules.import_code = () => ''
  }).use(codeSandboxPlugin)
}

describe('codeSandboxPlugin', () => {
  it('should work', () => {
    const md = createMarkdown()
    const code = `\
@[codesandbox](user/id)

@[codesandbox embed](user/id)

@[codesandbox button](user/id)

@[codesandbox title="xxx" layout="Editor+Preview" width="100%" height="500px" navbar="false" console="false"](user/slash#filepath)
`

    expect(md.render(code)).toMatchSnapshot()
  })

  it('should not work', () => {
    const md = createMarkdown()
    const code = `\
@[codesandbox]()

@[codesandbox]xxx

@[codesandbox embed](
`

    expect(md.render(code)).toMatchSnapshot()
  })
})
