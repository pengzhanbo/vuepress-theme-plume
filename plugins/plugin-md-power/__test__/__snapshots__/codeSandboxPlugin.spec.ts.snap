// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`codeSandboxPlugin > should not work 1`] = `
"<CodeSandboxViewer title="" height="500px" width="100%" user="" id="" type="embed" filepath="" :console=false :navbar=true layout="" /><p>@[codesandbox]xxx</p>
<p>@[codesandbox embed](</p>
"
`;

exports[`codeSandboxPlugin > should work 1`] = `"<CodeSandboxViewer title="" height="500px" width="100%" user="user" id="id" type="embed" filepath="" :console=false :navbar=true layout="" /><CodeSandboxViewer title="" height="500px" width="100%" user="user" id="id" type="embed" filepath="" :console=false :navbar=true layout="" /><CodeSandboxViewer title="" height="500px" width="100%" user="user" id="id" type="button" filepath="" :console=false :navbar=true layout="" /><CodeSandboxViewer title="xxx" height="500px" width="100%" user="user" id="slash" type="embed" filepath="filepath" :console=false :navbar=false layout="Editor+Preview" />"`;
